app/models/product.rb

```diff
@@ -101,6 +101,24 @@ def add_checklist_test
     end
   end
 
+  # This method does nothing more than attempt to cleanup a lot of data instead of making rails do it,
+  # since rails is really bad at cleaning up quickly. A similar method is implemented in the vendor
+  # model.
+  def destroy
+    tasks = Task.where(:product_test_id.in => product_test_ids)
+    records = Record.where(:test_id.in => product_test_ids)
+    record_ids = records.pluck(:_id)
+    calculation_results = HealthDataStandards::CQM::PatientCache.where(:'value.patient_id'.in => record_ids)
+
+    calculation_results.delete
+    records.delete
+    tasks.delete
+
```

Comment:
`Task`s have many `TestExecution`s, so we should probably delete them as well, when we delete both the vendor and the product